src/main/java/io/logz/jmx2graphite/JavaAgentClient.java

```diff
@@ -0,0 +1,183 @@
+        }
+    }
+
+    @Override
+    public List<MetricValue> getMetrics(List<MetricBean> beans) throws MBeanClientPollingFailure{
+
+        List<MetricValue> metricValues = Lists.newArrayList();
+
+        for (MetricBean metricBean : beans) {
+            try {
+                AttributeList attributeList = server.getAttributes(new ObjectName(metricBean.getName()),
+                                                                   metricBean.getAttributes().toArray(new String[0]));
+
+                Map<String, Object> attrValues = new HashMap<>(attributeList.size());
+                attributeList.asList().forEach((attr) ->
+                        attrValues.put(attr.getName(), attr.getValue()));
+
+                Map<String, Number> metricToValue = flatten(attrValues);
+
+                for (String attrMetricName : metricToValue.keySet()) {
+                    try {
+
+                        long metricTime = LocalDate.now().atStartOfDay(ZoneId.of("UTC")).toEpochSecond();
+
+                        metricValues.add(new MetricValue(
+                                GraphiteClient.sanitizeMetricName(metricBean.getName(), /*keepDot*/ true) + "." + attrMetricName,
+                                metricToValue.get(attrMetricName),
+                                metricTime));
+                    } catch (IllegalArgumentException e) {
+                        logger.info("Can't sent Metric since it's invalid: "+e.getMessage());
+                    }
+                }
+            } catch (MalformedObjectNameException | ReflectionException | InstanceNotFoundException | IllegalArgumentException e ) {
+                throw new MBeanClientPollingFailure(e.getMessage(), e);
+            }
+        }
+
+        return metricValues;
+    }
+
+    private Map<String, Number> flatten(Map<String, Object> attrValues) {
+
+        Map<String, Number> metricValues = Maps.newHashMap();
+        for (String key : attrValues.keySet()) {
+
+            Object value = attrValues.get(key);
+
+            if (value == null) {
+                continue;
+            }
+
+            if (value.getClass().isArray()) continue;
+            else if (value instanceof Number) {
+                metricValues.put(GraphiteClient.sanitizeMetricName(key, /*keepDot*/ false), (Number) value);
+            } else if (value instanceof CompositeData) {
+
+                CompositeData data = (CompositeData) value;
+                Map<String, Object> valueMap = handleCompositeData(data);
+                metricValues.putAll(prependKey(key, flatten(valueMap)));
+
+            } else if (value instanceof TabularData) {
+
+                TabularData tabularData = (TabularData) value;
+                Map<String, Object> rowKeyToRowData = handleTabularData(tabularData);
+                metricValues.putAll(prependKey(key, flatten(rowKeyToRowData)));
+
+            } else if (!(value instanceof String) && !(value instanceof Boolean)) {
```

Comment:
I just read Jolokia JVM Agent code and it appears they have extra protection against cases like infinit recursion, and also may have extra type handling.
One option I see is re-using this class they wrote "ObjectToJsonConverter". Let's talk about it